File: OP_PROJECT_JAVA/src/game/WorldPanel.java
package game;

import javax.swing.*;
import java.awt.*;
import java.awt.event.*;
import java.io.Serial;

public class WorldPanel extends JPanel {
    @Serial
    private static final long serialVersionUID = 1L;
    private static final int CELL_SIZE = 40;
    private static final int WORLD_SIZE_WIDTH = 20;
    private static final int WORLD_SIZE_HEIGHT = 11;
    private static final int INFO_HEIGHT = 176;
    private static final int BUTTONS_LEVEL = 470;
    private static final int BUTTON_HEIGHT = 40;
    private static final int BUTTON_WIDTH = 100;
    private final World world;
    private final JTextArea infoTextArea;

    public WorldPanel(int width, int height) {
        world = new World(WORLD_SIZE_WIDTH, WORLD_SIZE_HEIGHT, CELL_SIZE, this);

        setPreferredSize(new Dimension(width, height));
        setLayout(null);

        infoTextArea = new JTextArea();
        infoTextArea.setEditable(false);
        infoTextArea.setLineWrap(true);
        infoTextArea.setWrapStyleWord(true);

        JScrollPane scrollPane = new JScrollPane(infoTextArea);
        scrollPane.setVerticalScrollBarPolicy(JScrollPane.VERTICAL_SCROLLBAR_ALWAYS);
        scrollPane.setBounds(0, height - INFO_HEIGHT, width, INFO_HEIGHT);
        add(scrollPane); // Add the scroll pane to the panel

        // Create the "Next Turn" button
        JButton makeTurnButton = new JButton("Next Turn");
        makeTurnButton.setBounds(0, BUTTONS_LEVEL, BUTTON_WIDTH, BUTTON_HEIGHT);
        add(makeTurnButton);

        // Create the "Save Game" button
        JButton saveButton = new JButton("Save Game");
        saveButton.setBounds(BUTTON_WIDTH + 10, BUTTONS_LEVEL, BUTTON_WIDTH, BUTTON_HEIGHT);
        add(saveButton);

        // Create the "Load Game" button
        JButton loadButton = new JButton("Load Game");
        loadButton.setBounds(BUTTON_WIDTH * 2 + 20, BUTTONS_LEVEL, BUTTON_WIDTH, BUTTON_HEIGHT);
        add(loadButton);

        makeTurnButton.addActionListener(e -> {
            world.makeTurn();
            if (!world.isHumanAlive()) {
                endScreen();
            } else {
                repaint();
                requestFocusInWindow();
            }
        });

        saveButton.addActionListener(e -> {
            world.saveToFile();
            addTextToInfoPanel("Game state saved.");
            requestFocusInWindow();
        });

        loadButton.addActionListener(e -> {
            world.loadFromFile();
            addTextToInfoPanel("Game state loaded.");
            repaint();
            requestFocusInWindow();
        });

        addMouseListener(new MouseAdapter() {
            @Override
            public void mouseClicked(MouseEvent e) {
                int mouseX = e.getX() / CELL_SIZE;
                int mouseY = e.getY() / CELL_SIZE;
                if (world.isPositionWithinBounds(new Position(mouseY, mouseX))) {
                    Organism existingOrganism = world.getOrganismAtPosition(new Position(mouseY, mouseX));
                    if (existingOrganism == null) {
                        String[] options = {"Wolf", "Sheep", "Fox", "Turtle", "Antelope", "Grass", "Sow Thistle", "Guarana", "Belladonna", "Sosnowsky Hogweed"};
                        String chosenType = (String) JOptionPane.showInputDialog(
                                null,
                                "Choose the type of organism to add:",
                                "Add Organism",
                                JOptionPane.PLAIN_MESSAGE,
                                null,
                                options,
                                options[0]);

                        if (chosenType != null) {
                            // Add the new organism at the clicked position
                            world.bornAnimal(chosenType, new Position(mouseY, mouseX));
                            repaint();
                        }
                    }
                }
            }
        });

        addKeyListener(new KeyAdapter() {
            @Override
            public void keyPressed(KeyEvent e) {
                handleInput(e);
            }
        });

        setFocusable(true);
        requestFocusInWindow();

    }

    public void handleInput(KeyEvent e) {
        char ch = e.getKeyChar();
        switch (ch) {
            case 'w':
            case 's':
            case 'a':
            case 'd':
                world.setNextMove(ch);
                break;
            case 'p': // Special power of human
                world.activateSpecialPower();
                break;
        }
    }

    public void disableInput() {
        // Remove mouse click listener
        for (MouseListener listener : this.getMouseListeners()) {
            this.removeMouseListener(listener);
        }

        // Remove key listener
        for (KeyListener listener : this.getKeyListeners()) {
            this.removeKeyListener(listener);
        }

        // Remove all buttons
        Component[] components = this.getComponents();
        for (Component component : components) {
            if (component instanceof JButton) {
                this.remove(component);
            }
        }

        // Remove the scroll pane
        for (Component component : components) {
            if (component instanceof JScrollPane) {
                this.remove(component);
            }
        }
    }

    private void endScreen() {
        clearScreen();
        drawText("YOU LOST", getWidth() / 2 - 50, getHeight() / 2);
        disableInput();
    }

    public void clearScreen() {
        Graphics2D g2d = (Graphics2D) getGraphics();
        g2d.clearRect(0, 0, getWidth(), getHeight());
        g2d.dispose();
    }

    public void drawText(String text, int x, int y) {
        Graphics2D g2d = (Graphics2D) getGraphics();
        g2d.setColor(Color.RED);
        g2d.setFont(new Font("Arial", Font.BOLD, 24));
        g2d.drawString(text, x, y);
        g2d.dispose();
    }

    @Override
    protected void paintComponent(Graphics g) {
        super.paintComponent(g);
        Graphics2D g2d = (Graphics2D) g.create();
        world.draw(g2d);
        g2d.dispose();
    }

    public void addTextToInfoPanel(String text) {
        infoTextArea.append(text + "\n");
        infoTextArea.setCaretPosition(infoTextArea.getDocument().getLength());
    }
}

File: OP_PROJECT_PYTHON/src/game/WorldPanel.py
import tkinter as tk
from tkinter import ttk, messagebox
import pickle
from Position import Position
from World import World


class WorldPanel(tk.Frame):
    CELL_SIZE = 40
    WORLD_SIZE_WIDTH = 20
    WORLD_SIZE_HEIGHT = 11
    INFO_HEIGHT = 180
    BUTTON_HEIGHT = 40
    BUTTON_WIDTH = 100

    def __init__(self, master, width, height, *args, **kwargs):
        super().__init__(master, *args, **kwargs)

        self.world = World(self.WORLD_SIZE_WIDTH, self.WORLD_SIZE_HEIGHT, self.CELL_SIZE, self)

        self.grid_panel = tk.Canvas(self, width=width, height=height - self.INFO_HEIGHT - self.BUTTON_HEIGHT, bg="white")
        self.grid_panel.bind("<Button-1>", self.on_canvas_click)

        self.info_text_area = tk.Text(self, wrap=tk.WORD)
        self.info_text_area.config(state=tk.DISABLED)
        self.scrollbar = tk.Scrollbar(self, command=self.info_text_area.yview)
        self.info_text_area.config(yscrollcommand=self.scrollbar.set)

        self.make_turn_button = tk.Button(self, text="Next Turn", command=self.make_turn)
        self.save_button = tk.Button(self, text="Save Game", command=self.save_game)
        self.load_button = tk.Button(self, text="Load Game", command=self.load_game)

        self.organism_options = ["Wolf", "Sheep", "Fox", "Turtle", "Antelope", "Grass", "Sow Thistle", "Guarana", "Belladonna", "Sosnowsky Hogweed", "Cyber Sheep"]

        self.setup_ui()
        self.bind_all("<Key>", self.handle_input)

        self.world.draw(self.grid_panel)
        self.grid_panel.update()

    def setup_ui(self):
        self.grid_panel.pack(side=tk.TOP, fill=tk.BOTH, expand=True)

        self.make_turn_button.pack(side=tk.LEFT, padx=5, pady=5)
        self.save_button.pack(side=tk.LEFT, padx=5, pady=5)
        self.load_button.pack(side=tk.LEFT, padx=5, pady=5)

        self.info_text_area.pack(side=tk.LEFT, fill=tk.BOTH, expand=True)
        self.scrollbar.pack(side=tk.RIGHT, fill=tk.Y)

    def make_turn(self):
        self.world.make_turn()
        if not self.world.is_human_alive():
            self.end_screen()
        else:
            self.grid_panel.delete("all")
            self.world.draw(self.grid_panel)
            self.grid_panel.update()

    def save_game(self):
        try:
            with open('savefile.pkl', 'wb') as f:
                pickle.dump(self.world, f)
            messagebox.showinfo("Success", "Game state saved.")
            self.focus_set()
        except Exception as ex:
            print(ex)
            messagebox.showerror("Error", f"Failed to save the game: {ex}")

    def load_game(self):
        try:
            with open('savefile.pkl', 'rb') as f:
                loaded_world = pickle.load(f)
                # Update the current world instance with the loaded one
                self.world.__dict__.update(loaded_world.__dict__)
                # Reassign the world reference for each organism
                for organism in self.world.organisms:
                    organism.world = self.world
                self.grid_panel.delete("all")
                self.world.draw(self.grid_panel)
                self.grid_panel.update()
            messagebox.showinfo("Success", "Game loaded successfully")
        except Exception as e:
            messagebox.showerror("Error", "Failed to load the game")

    def handle_input(self, event):
        ch = event.char
        if ch in ['w', 's', 'a', 'd']:
            self.world.set_next_move(ch)
        elif ch == 'p':
            self.world.activate_special_power()

    def on_canvas_click(self, event):
        mouse_x = event.x // self.CELL_SIZE
        mouse_y = event.y // self.CELL_SIZE
        if self.world.is_position_within_bounds(Position(mouse_y, mouse_x)):
            existing_organism = self.world.get_organism_at_position(Position(mouse_y, mouse_x))
            if existing_organism is None:
                self.open_organism_selection_window(Position(mouse_y, mouse_x))

    def open_organism_selection_window(self, position):
        self.selection_window = tk.Toplevel(self)
        self.selection_window.title("Choose Organism")

        # Center the selection window relative to the main window
        window_width = 300
        window_height = 150
        screen_width = self.master.winfo_screenwidth()
        screen_height = self.master.winfo_screenheight()
        x_cordinate = int((screen_width / 2) - (window_width / 2))
        y_cordinate = int((screen_height / 2) - (window_height / 2))
        self.selection_window.geometry(f"{window_width}x{window_height}+{x_cordinate}+{y_cordinate}")

        self.organism_var = tk.StringVar(self.selection_window)
        self.organism_var.set(self.organism_options[0])  # default value

        label = tk.Label(self.selection_window, text="Choose the type of organism to add:")
        label.pack(pady=10)

        self.organism_menu = ttk.Combobox(self.selection_window, textvariable=self.organism_var, values=self.organism_options, state='readonly')
        self.organism_menu.pack(pady=10)

        add_button = tk.Button(self.selection_window, text="Add Organism", command=lambda: self.add_organism(position))
        add_button.pack(pady=10)

    def add_organism(self, position):
        chosen_type = self.organism_var.get()
        if chosen_type in self.organism_options:
            self.world.born_organism(chosen_type, position)
            self.world.draw(self.grid_panel)
            self.grid_panel.update()
        self.selection_window.destroy()

    def add_text_to_info_panel(self, text):
        self.info_text_area.config(state=tk.NORMAL)
        self.info_text_area.insert(tk.END, text + "\n")
        self.info_text_area.see(tk.END)
        self.info_text_area.config(state=tk.DISABLED)

    def end_screen(self):
        self.disable_input()
        self.grid_panel.delete("all")
        self.grid_panel.create_text(self.grid_panel.winfo_width() / 2, self.grid_panel.winfo_height() / 2, text="YOU LOST", fill="red", font=("Arial", 24))

    def disable_input(self):
        self.grid_panel.unbind("<Button-1>")
        self.unbind_all("<Key>")
        for widget in [self.make_turn_button, self.save_button, self.load_button]:
            widget.config(state=tk.DISABLED)
